### PR TITLE
StateMachine.getOnChainState throws NoOnChainState

### PR DESCRIPTION
**Pull request by nau**
Marlowe and other contracts can re-use `getOnChainState` and handle the case when no onchain state is currently present.

### CODE REVIEW

File: plutus-contract/src/Language/Plutus/Contract/StateMachine.hs [hunk 1]
```diff
@@ -198,8 +201,9 @@ waitForUpdateUntil StateMachineClient{scInstance, scChooser} timeoutSlot = do
                 then go (succ sl)
                 else pure txns
 
+    initial <- currentSlot
+    txns <- go initial
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:205]: Why this change?

File: plutus-contract/src/Language/Plutus/Contract/StateMachine.hs [hunk 2]
```diff
@@ -161,7 +162,9 @@ getOnChainState ::
 getOnChainState StateMachineClient{scInstance, scChooser} = mapError (review _SMContractError) $ do
     utxo <- utxoAt (SM.machineAddress scInstance)
     let states = getStates scInstance utxo
-    either (throwing _SMContractError) (\s -> pure (s, utxo)) (scChooser states)
+    case states of
+        [] -> throwError NoOnChainState
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:166]: Maybe we should just return a maybe here? Arguably this is more of a "normal" return than an error? Not sure.

File: plutus-contract/src/Language/Plutus/Contract/StateMachine.hs (hunk 1 above)
Comment (nau) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:205]: It was a bug. I used this `initial slot` to compare with `timeout` slot in `case states of`, and it could not ever `Timeout`.

File: plutus-contract/src/Language/Plutus/Contract/StateMachine.hs (hunk 2 above)
Comment (nau) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:166]: Not sure either. It's sort of handling a special case, not all contracts would want to handle it.

File: plutus-contract/src/Language/Plutus/Contract/StateMachine.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:205]: Sounds like it deserves a comment!

File: plutus-contract/src/Language/Plutus/Contract/StateMachine.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:166]: So I guess the other errors aren't really "recoverable" ones. I think in this case it would be sensible to return a maybe - the user can just throw immediately if they can't handle that.

Comment (michaelpj) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:166]: And arguably "one participant starts trying to do things before the machine has started" is actually a very common case that users *should* handle. Making it an error makes it easy to just "let it crash" in this case, which probably isn't right.

Comment (nau) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:166]: Fair enough.

**Review by j-mueller — APPROVED** [review]
LGTM

File: plutus-contract/src/Language/Plutus/Contract/StateMachine.hs [hunk 3]
```diff
@@ -333,7 +336,8 @@ mkStep ::
     -> Contract schema e (StateMachineTransition state input)
 mkStep client@StateMachineClient{scInstance} input = do
     let StateMachineInstance{stateMachine=StateMachine{smTransition}, validatorInstance} = scInstance
-    (onChainState, utxo) <- getOnChainState client
+    maybeState <- getOnChainState client
+    (onChainState, utxo) <- maybe (throwing_ _NoOnChainState) pure maybeState
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:340]: Maybe we should propagate it upwards here too... not sure.

Comment (michaelpj) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:340]: Relevant to Jann's open PR.

Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs:340]: I ended up removing the `NoOnChainState` failure in https://github.com/input-output-hk/plutus/pull/2511. There is now a type

```haskell
data InvalidTransition = InvalidTransition
        { tfState :: Maybe (State s)
        , tfInput :: i
        }
```

where `tfState = Nothing` is used for the case that no on-chain state was found.
